sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,788 @@
+    abstract String getQuery();
+
+    abstract String getScrollKeepalive();
+
+    abstract long getBatchSize();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+
+      abstract Builder setQuery(String query);
+
+      abstract Builder setScrollKeepalive(String scrollKeepalive);
+
+      abstract Builder setBatchSize(long batchSize);
+
+      abstract Read build();
+    }
+
+    /**
+     * Provide the Elasticsearch connection configuration object.
+     *
+     * @param connectionConfiguration the Elasticsearch {@link ConnectionConfiguration} object
+     * @return the {@link Read} with connection configuration set
+     */
+    public Read withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(
+          connectionConfiguration != null,
+          "ElasticsearchIO.read()"
+              + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    /**
+     * Provide a query used while reading from Elasticsearch.
+     *
+     * @param query the query. See <a
+     *     href="https://www.elastic.co/guide/en/elasticsearch/reference/2.4/query-dsl.html">Query
+     *     DSL</a>
+     * @return the {@link Read} object with query set
+     */
+    public Read withQuery(String query) {
+      checkArgument(
+          query != null, "ElasticsearchIO.read().withQuery(query) called with null query");
+      return builder().setQuery(query).build();
+    }
+
+    /**
+     * Provide a scroll keepalive. See <a
+     * href="https://www.elastic.co/guide/en/elasticsearch/reference/2.4/search-request-scroll.html">scroll
+     * API</a> Default is "5m". Change this only if you get "No search context found" errors.
+     *
+     * @param scrollKeepalive keepalive duration ex "5m" from 5 minutes
+     * @return the {@link Read} with scroll keepalive set
+     */
+    public Read withScrollKeepalive(String scrollKeepalive) {
+      checkArgument(
+          scrollKeepalive != null,
+          "ElasticsearchIO.read().withScrollKeepalive(keepalive) called with null keepalive");
+      return builder().setScrollKeepalive(scrollKeepalive).build();
+    }
+
+    /**
+     * Provide a size for the scroll read. See
+     * <a href="https://www.elastic.co/guide/en/elasticsearch/reference/2.4/search-request-scroll.html">
+     *   scroll API</a> Default is 100. Maximum is 10 000. If documents are small,
+     *   increasing batch size might improve read performance. If documents are big, you might need
+     *   to decrease batchSize
+     *
+     * @param batchSize number of documents read in each scroll read
+     * @return the {@link Read} with batch size set
+     */
+    public Read withBatchSize(long batchSize) {
+      return builder().setBatchSize(
```

Comment:
I have no reply button on previous comment, so I'll answer in this one:
Yes, actually, I put IT material in the PR to launch the discussion about integration tests. There is an ongoing discussion on the ML with JB and Stephen on where to put IT material (separate IT module or IO module). The script you mention is to be launched at backend instance creation time (it could be run only once or before running the IT test suite if we decide to recreate instances each time before running tests suites). As external ES instance is needed for the IT tests to run (and logstash ingestion also), then, yes, this script is needed for IT tests to run.

For above comment on withBatchSize: I aggree